Implement Win64 eh_personality natively.

vadimcn opened the pull request:
After this change, the only remaining symbol we are pulling from libgcc on Win64 is `__chkstk_ms` - the stack probing routine.

rust-highfive commented:
r? @brson

(rust_highfive has picked a reviewer for you, use r? to override)

vadimcn commented:
I am not quite happy about the way `rust_eh_unwind_resume` get linked.  It should have probably been a lang item, but I couldn't quite figure out how to plumb it through to `LLVMRustHookUnwindResume`...

r? @alexcrichton

vadimcn commented:
r? @alexcrichton

bors commented:
:umbrella: The latest upstream changes (presumably #27176) made this pull request unmergeable. Please resolve the merge conflicts.

nagisa commented on src/doc/trpl/no-stdlib.md:
```diff
@@ -71,18 +73,18 @@ The compiler currently makes a few assumptions about symbols which are available
-[libstd implementation](../std/rt/unwind/index.html) for more
+The second and third of these functions, `eh_personality` and `rust_eh_unwind_resume`,
+are used by the failure mechanisms of the compiler. These are often mapped to GCC's
+personality function and libgcc's `_Unwind_Resume` respectively
```
Wording of this sentence somehow seems to suggest GCC personality function and `_Unwind_Resume` come from two different locations, while in practice they both come from `libgcc`.

nagisa commented on src/librustc_trans/back/write.rs:
```diff
@@ -512,6 +517,13 @@ unsafe fn optimize_and_codegen(cgcx: &CodegenContext,
 
+    // Inject a local-scoped `_Unwind_Resume` function, which redirects to `rust_eh_unwind_resume`.
+    // This must be done after optimizations, because until DwarfEHPrepare pass has been run,
+    // `_Unwind_Resume` is not referenced by any live code and would have been eliminated by opt.
```
Wouldn’t [OptimizeNone](http://llvm.org/docs/doxygen/html/Attributes_8h_source.html#l00098) prevent it from being eliminated?

alexcrichton replied:
This may also be able to prevent elimination through the [`llvm.used` global](http://llvm.org/docs/LangRef.html#the-llvm-used-global-variable)

alexcrichton commented on src/doc/trpl/no-stdlib.md:
```diff
@@ -38,6 +38,7 @@ fn start(_argc: isize, _argv: *const *const u8) -> isize {
 // provided by libstd.
 #[lang = "stack_exhausted"] extern fn stack_exhausted() {}
 #[lang = "eh_personality"] extern fn eh_personality() {}
+#[no_mangle] pub extern fn rust_eh_unwind_resume() {}
```
It looks like this only needs to be defined on 64-bit windows targets, so can this be omitted? I suspect 99% of `#![no_std]` users aren't working on Windows.

vadimcn replied:
Without it doctests fail on Windows.  Also, if we ever drop libgcc on other platforms, we'll need it.

alexcrichton replied:
Can you hide this by default? (e.g. put a `#` in front)

alexcrichton commented on src/librustc_back/target/x86_64_pc_windows_gnu.rs:
```diff
@@ -16,6 +16,7 @@ pub fn target() -> Target {
     // On Win64 unwinding is handled by the OS, so we can link libgcc statically.
     base.pre_link_args.push("-static-libgcc".to_string());
     base.pre_link_args.push("-m64".to_string());
+    base.custom_unwind_resume = true;
```
Should this also set the hook for x86_64-pc-windows-msvc?

vadimcn replied:
I was thinking of a separate PR.

alexcrichton replied:
Ok, sounds good to me

alexcrichton commented on src/libstd/rt/dwarf/eh.rs:
```diff
@@ -0,0 +1,141 @@
+}
+
+#[inline(never)]
+pub unsafe fn find_landing_pad(lsda: *const u8, context: EHContext) -> Option<usize> {
```
A few notes on this function:

* Is there public documentation available for this protocol? I'd love to reference this against a known implementation or description to verify it, but also so we can understand what this is doing months from now.
* Can you document why this is `inline(never)`?

vadimcn replied:
> Is there public documentation available for this protocol? 

Not really.  Gcc LSDA format is defined by, well, gcc implementation.  Aside from that, there are a few blog posts...  I'll link them here.

vadimcn replied:
inline(never) isn't needed, just forgot to delete it

alexcrichton replied:
Yeah as many links as possible would be nice, even if they're just links to code in other repos. Surely this implementation came from somewhere, and it'd be good to have a trail saying how so!

alexcrichton commented on src/libstd/rt/dwarf/mod.rs:
```diff
@@ -0,0 +1,97 @@
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! General DWARF format parsing utilities
```
Similarly as above, it'd be great to have a lot more documentation on this module. For example the `read_uleb128` function would be nice to have a hyperlink to the definition of the format, and some overall docs about what this module's doing would be nice.

vadimcn replied:
Now these are defined in the DWARF standard.  I'll add a section reference here.

alexcrichton commented on src/libstd/rt/unwind/gcc.rs:
```diff
@@ -129,7 +120,10 @@ pub mod eabi {
+            unsafe {
+                __gcc_personality_v0(version, actions, exception_class, ue_header,
+                                     context)
+            }
```
Just to clarify, was this a drive-by fix?

alexcrichton replied:
Oh hm ok I think I see what's going on here, it's because this is legit running in something other than the search phase, so we probably need to do something more than just returning a constant.

vadimcn replied:
This is basically a roll-back of the rust_try_inner magic, which was created because of the way libgcc's C personality routine works on win64.  I thought it'd be good to get rid of unneeded complications. This is the way this code used to be a year ago.

alexcrichton replied:
Right, thanks!

vadimcn replied:
I'll try to make the equivalent simplification in your `trans_gnu_try`.

alexcrichton commented on src/libstd/rt/unwind/gcc.rs:
```diff
@@ -99,8 +92,6 @@ pub mod eabi {
 
     #[lang="eh_personality"]
-    #[no_mangle] // referenced from rust_try.ll
-    #[allow(private_no_mangle_fns)]
```
Does this pass all the tests? I vaguely remember seeing the `rust_eh_personality` symbol in a list of reachable symbols which was necessary for LTO working, but I may be remembering incorrectly.

vadimcn replied:
It did, on Windows.  Might be worth pushing this into try?

alexcrichton replied:
Ah well, if it passes all the bots then it's fine by me!

alexcrichton commented on src/libstd/rt/mod.rs:
```diff
@@ -47,6 +47,9 @@ pub mod args;
 mod at_exit_imp;
 mod libunwind;
 
+#[cfg(all(windows, target_arch="x86_64", target_env="gnu"))]
```
It may be worth having this just be compiled for all platforms and it'll just end up being optimized away on most. That way we can catch breakage to it on all platforms instead of just 64-bit windows.

alexcrichton commented on src/libstd/rt/unwind/seh64_gnu.rs:
```diff
@@ -0,0 +1,231 @@
+// Since these assumptions do not generally hold true for foreign exceptions (system faults,
+// C++ exceptions, etc), we make no attempt to invoke our landing pads (and, thus, destructors!)
+// for anything other than RUST_PANICs. This is considered acceptable, because the behavior of
+// throwing exceptions through a C ABI boundary is undefined.
```
Could you format this to 80-chars wide? (what the rest of the runtime is formatted to)

alexcrichton commented on src/libstd/rt/unwind/seh64_gnu.rs:
```diff
@@ -0,0 +1,231 @@
+
+    if er.ExceptionFlags & EXCEPTION_UNWIND == 0 { // we are in the dispatch phase
+        if er.ExceptionCode == RUST_PANIC {
+            if let Some(lpad) = find_landing_pad(dc) {
```
This may be a bit clearer as:

```
let lpad = find_landing_pad(dc).unwrap_or_else(|| {
    rtabort!(...)
});
```

alexcrichton commented on src/libstd/rt/unwind/seh64_gnu.rs:
```diff
@@ -0,0 +1,231 @@
+
+#[lang="eh_personality"]
+#[cfg(not(test))]
+unsafe extern "C" fn rust_eh_personality(
```
Is this just the same as `rust_eh_personality_catch`? These two functions look the same except for the expectation of the landing pad existing...

Overall I'd find it at least helpful to have a nice big doc comment at the top of the module explaining what's going on here. I have a little understanding of what mechanisms are in play, but I know why these are structured the way they are. For example it's not clear to my why the catch personality is catching but this one isn't?

vadimcn replied:
> Is this just the same as rust_eh_personality_catch? These two functions look the same except for the expectation of the landing pad existing...

Yes, that the only difference at the moment.  There used to be more when I tried to support running destructors on foreign exceptions.

alexcrichton replied:
Interesting... I guess this shows how little I know about the ABI of these exceptions on windows! I guess whether a landing pad is a "catch" landing pad is encoded in the landing pad itself and not the personality function?

Overall it basically may mean that we can merge these two functions and only have one?

alexcrichton commented on src/libstd/rt/unwind/seh64_gnu.rs:
```diff
@@ -0,0 +1,231 @@
+    ExceptionContinueSearch
+}
+
+#[no_mangle] // referenced from landing pads
```
Can you expand this to a comment explaining what this function is doing? It'd be nice to mention that this is called from all landing pads and is injected by LLVM, and it's also why it's `pub` + `#[no_mangle]`.

Also, can you remove the `"C"` part from the header?

alexcrichton commented:
Nice work @vadimcn!

liigo commented on src/libstd/rt/unwind/seh64_gnu.rs:
```diff
@@ -0,0 +1,231 @@
+//    [24:27] = type
+//    [0:23]  = magic
+const ETYPE: DWORD = 0b1110_u32 << 28;
+const MAGIC: DWORD = 0x525354; // "RST"
```
`RST` also refers to `reStructuredText`. How about we use `.rs` here?
```
const MAGIC: DWORD = 0x2e7273; // ".rs"
```

brson commented:
Epic patch.

vadimcn commented:
Addressed comments and rebased on top of master. r?

alexcrichton commented on src/libstd/rt/unwind/mod.rs:
```diff
@@ -76,9 +76,20 @@ use sys_common::mutex::Mutex;
 pub mod imp;
-#[cfg(not(target_env = "msvc"))] #[path = "gcc.rs"] #[doc(hidden)]
+
+// x86-pc-windows-gnu and all others
```
s/x86/i686/

alexcrichton commented on src/libstd/rt/unwind/seh64_gnu.rs:
```diff
@@ -0,0 +1,248 @@
+// `_Unwind_Resume` routine.  To avoid confusion with the same symbol exported
+// from libgcc, we redirect it to `rust_eh_unwind_resume`.
+// Since resolution of this symbol is done by the linker, `rust_eh_unwind_resume`
+// must be marked `pub` + `#[no_mangle]`.  (Can we make it a lang item?)
```
Unfortunately I don't think this can be a lang item because libraries like libcore/libcollections need to reference this (they're generating landing pads) so it needs to have a well known symbol name (e.g. it's used before it's defined).

alexcrichton commented on src/libstd/rt/unwind/seh64_gnu.rs:
```diff
@@ -0,0 +1,248 @@
+    if er.ExceptionFlags & EXCEPTION_UNWIND == 0 { // we are in the dispatch phase
+        if er.ExceptionCode == RUST_PANIC {
+            if let Some(lpad) = find_landing_pad(dc) {
+                RtlUnwindEx(establisherFrame,
```
So if I'm understanding this correctly, this call to `RtlUnwindEx` magically jumps back into the function and executes some code? The [documentation](https://msdn.microsoft.com/en-us/library/windows/desktop/ms680615%28v=vs.85%29.aspx) is a bit scant, and I thought when poking through the libgcc sources that this was basically re-raising the exception, but I also have no idea what's going on here :)

Also, could the `eh_personality` and `eh_personality_catch` functions be merged? Or could the catch one just call the normal one?

alexcrichton replied:
(as in the extra assertion in the catch that a landing pad is there doesn't seem worth the duplication)

vadimcn replied:
> So if I'm understanding this correctly, this call to RtlUnwindEx magically jumps back into the function and executes some code?

Yes, pretty much.  It unwinds stack down to that function's frame, then resumes execution at the specified IP.   The way I chose to implement this is a bit different than what gcc does; their way ended up too complicated, IMO, because they tried to re-use the same personality routine used for DWARF-style unwinding.

> Also, could the eh_personality and eh_personality_catch functions be merged?

I kinda prefer to keep them separate,- for symmetry with other unwinding implementations.  But OK, I'll merge them.

alexcrichton replied:
Oh I think it's fine to keep two personality functions, I just figured it'd be good to reduce the code duplication.

alexcrichton replied:
(ping on this)

vadimcn replied:
Okay, if you insist...

alexcrichton commented on src/rustllvm/RustWrapper.cpp:
```diff
@@ -958,3 +958,27 @@ LLVMRustBuildLandingPad(LLVMBuilderRef Builder,
+    PointerType *elem_type = PointerType::get(IntegerType::get(Ctx, 8), 0);
+    ArrayType *used_type = ArrayType::get(elem_type, 1);
+    Constant* init = ConstantArray::get(used_type, ArrayRef<Constant*>(ConstantExpr::getBitCast(OrigRewind, elem_type)));
+    GlobalVariable* used = new GlobalVariable(*M, used_type, false, GlobalValue::AppendingLinkage, init, "llvm.used");
```
Could you be sure to wrap these lines to 80 chars as well?

alexcrichton commented on src/rustllvm/RustWrapper.cpp:
```diff
@@ -958,3 +958,27 @@ LLVMRustBuildLandingPad(LLVMBuilderRef Builder,
+    Constant* init = ConstantArray::get(used_type, ArrayRef<Constant*>(ConstantExpr::getBitCast(OrigRewind, elem_type)));
+    GlobalVariable* used = new GlobalVariable(*M, used_type, false, GlobalValue::AppendingLinkage, init, "llvm.used");
+    used->setSection("llvm.metadata");
+}
```
Can this all be written in Rust? It's generally quite difficult to keep C++ up to date across LLVM versions and the C API is much more stable.

vadimcn replied:
Hm, ok.  Never thought about it from this angle.

alexcrichton commented on src/libstd/rt/dwarf/mod.rs:
```diff
@@ -0,0 +1,107 @@
+        let mut byte : u8;
+        loop {
+            byte = *self.ptr;
+            self.ptr = self.ptr.offset(1);
```
Perhaps this could be `let byte = self.read::<u8>()` to consolidate the shifting of `ptr`?

alexcrichton commented on src/libstd/rt/dwarf/mod.rs:
```diff
@@ -0,0 +1,107 @@
+//! See http://www.dwarfstd.org,
+//! DWARF-4 standard, Section 7 - "Data Representation"
+
+#![allow(unused)]
```
Could you add a small comment to this indicating that this module isn't used on all platforms?

alexcrichton commented on src/libstd/rt/dwarf/eh.rs:
```diff
@@ -0,0 +1,150 @@
+}
+
+unsafe fn read_encoded_pointer(reader: &mut DwarfReader,
+                               context: EHContext,
```
Perhaps this could be passed by reference? It looks like this doesn't necessarily need to take ownershipo

alexcrichton commented on src/libstd/rt/dwarf/eh.rs:
```diff
@@ -0,0 +1,150 @@
+            if cs_lpad != 0 {
+                return Some(lpad_base + cs_lpad);
+            } else {
+                return None;
```
Should this and the "not found" case be differentiated? http://www.airs.com/blog/archives/464 seems to indicate that C++ aborts in the "not found" case but simply continues in the case of "`cs_lpad` is 0". May be worth having the extra assertion on the other end?

vadimcn replied:
Right.   But the [libgcc's implementation](https://github.com/gcc-mirror/gcc/blob/master/libgcc/unwind-c.c#L216) actually treats them the same.  Not sure which one is correct.  Not emitting info for an IP range, if no action needs to be taken, could be a valid optimization to reduce LSDA size, IMO.

alexcrichton replied:
Hm ok, could you add a comment to that effect?

alexcrichton commented on src/libstd/rt/dwarf/eh.rs:
```diff
@@ -0,0 +1,150 @@
+        let cs_start = read_encoded_pointer(&mut reader, context, call_site_encoding);
+        let cs_len = read_encoded_pointer(&mut reader, context, call_site_encoding);
+        let cs_lpad = read_encoded_pointer(&mut reader, context, call_site_encoding);
+        let cs_action = reader.read_uleb128();
```
Should we have an assertion that this is 0 most of the time? It sounds like the only nonzero entry should be the catch block, right?

alexcrichton replied:
(ping on this)

vadimcn replied:
ok

vadimcn replied:
Actually, this is not true - the action for `rust_try`'s landing pad is not 0, though it does point to a null entry in the types table.   I think I'd rather not implement exception type lookup logic just to assert that it's a "catch-all".

alexcrichton commented on src/libstd/rt/dwarf/eh.rs:
```diff
@@ -0,0 +1,150 @@
+        _ => panic!()
+    };
+
+    result += match encoding & 0x70 {
```
Should this mask be `0xF0`? It looks like this misses the `DW_EH_PE_indirect` case (although it may not actually come up in practice)

alexcrichton replied:
Oh I see you handle this case below. Perhaps it could be added to this `match` though?

vadimcn replied:
DW_EH_PE_indirect may be combined with any of the above flags, so it's less code this way.

alexcrichton replied:
Ah ok that makes sense

alexcrichton commented on src/libstd/rt/dwarf/eh.rs:
```diff
@@ -0,0 +1,150 @@
+
+    result += match encoding & 0x70 {
+        DW_EH_PE_absptr => 0,
+        // relative to address of the encoded value, despite the name
```
Well if that's not confusing I don't know what is!

alexcrichton commented on src/libstd/rt/unwind/seh64_gnu.rs:
```diff
@@ -0,0 +1,248 @@
+        ip: dc.ControlPc as usize,
+        func_start: dc.ImageBase as usize + (*dc.FunctionEntry).BeginAddress as usize,
+        text_start: dc.ImageBase as usize,
+        data_start: 0
```
Perhaps this field could be omitted for now? The `datarel` encoding may want to just panic saying "unimplemented".

alexcrichton commented on src/libstd/rt/dwarf/eh.rs:
```diff
@@ -0,0 +1,150 @@
+    let call_site_encoding = reader.read::<u8>();
+    let call_site_table_length = reader.read_uleb128();
+    let action_table = reader.ptr.offset(call_site_table_length as isize);
+    let ip = context.ip-1;
```
Could you add a comment to why this `-1` is done here?

alexcrichton commented:
After going through all the nitty gritty I'm even more impressed now, very nice work @vadimcn!

vadimcn commented:
Addressed review comments.

Also, now that I'm using `llvm.used` to prevent elimination of `_Unwind_Resume` hook by the optimizer, I moved hook injection earlier in trans, and was able to make `rust_eh_unwind_resume` a lang item!

alexcrichton commented on src/librustc_trans/trans/common.rs:
```diff
@@ -530,6 +530,51 @@ impl<'a, 'tcx> FunctionContext<'a, 'tcx> {
         }
     }
+
+    pub fn inject_unwind_resume_hook(&self) {
```
Could you add a doc comment here explaining what this function is doing?

alexcrichton commented on src/librustc_trans/trans/intrinsic.rs:
```diff
@@ -1188,108 +1176,61 @@ fn trans_gnu_try<'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
         let (fcx, block_arena);
         block_arena = TypedArena::new();
-        fcx = new_fn_ctxt(ccx, llfn, ast::DUMMY_NODE_ID, false,
+        fcx = new_fn_ctxt(ccx, rust_try, ast::DUMMY_NODE_ID, false,
```
If you're feeling ambitious it looks like the `_Unwind_Resume` shim you created above is much nicer to use than creating all these contexts here, so it may be easier to do that here as well. That being said it's just cleanup so feel free to pass!

bors commented:
:umbrella: The latest upstream changes (presumably #27309) made this pull request unmergeable. Please resolve the merge conflicts.

vadimcn commented:
looks ok now?

alexcrichton commented:
Looks good to me! r=me with a rebase

vadimcn commented:
rebased. r?

alexcrichton commented:
@bors: r+ aaa95a072b3ec8572fd5b2fe2f9409b494fb263a

bors commented:
:hourglass: Testing commit aaa95a0 with merge 698d222...

bors commented:
:broken_heart: Test failed - [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/884)

vadimcn commented:
Broken by last rebase :(

vadimcn commented:
Works now.  r?

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit e493027 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton e493027984b5961617c867a7003face00a057c92 -->

bors commented:
:hourglass: Testing commit e493027 with merge 2c60a05...

bors commented:
:broken_heart: Test failed - [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/5916)

Gankro commented:
Seems legit:

```

../src/libstd/rt/unwind/gcc.rs:231:38: 231:43 error: use of moved
value: `state` [E0382]
../src/libstd/rt/unwind/gcc.rs:231
__gcc_personality_v0(state, ue_header, context)
                                                                        ^~~~~
../src/libstd/rt/unwind/gcc.rs:225:13: 225:18 note: `state` moved here
because it has type `rt::libunwind::_Unwind_State`, which is
non-copyable
../src/libstd/rt/unwind/gcc.rs:225         if (state as c_int &
uw::_US_ACTION_MASK as c_int)
                                               ^~~~~
error: aborting due to previous error
make: *** [x86_64-unknown-linux-gnu/stage2/lib/rustlib/arm-linux-androideabi/lib/stamp.std]
Error 101
make: *** Waiting for unfinished jobs....

```

On Sun, Aug 2, 2015 at 12:08 PM, bors <notifications@github.com> wrote:

> [image: :broken_heart:] Test failed - auto-linux-64-x-android-t
> <http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/5916>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/27210#issuecomment-127055917>.
>

vadimcn commented:
Fixed.  Let's try again?

alexcrichton commented:
@bors: r+ 96d1db2

bors commented:
:hourglass: Testing commit 96d1db2 with merge d99b3c5...

bors commented:
:broken_heart: Test failed - [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/5957)

alexcrichton commented:
@bors: retry

On Mon, Aug 3, 2015 at 10:37 AM, bors <notifications@github.com> wrote:

> [image: :broken_heart:] Test failed - auto-mac-64-opt
> <http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/5957>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/27210#issuecomment-127341324>.
>

bors commented:
:hourglass: Testing commit 96d1db2 with merge ceded6a...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/5970), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/5985), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/5927), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/5955), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/5932), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/6006), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/5998), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/5963), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/943), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/946), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/942), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/945), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/236), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/932)

alexcrichton commented:
:confetti_ball:
